File: geekbang/algorithm-models/7.1- Bellman-Ford 求最短路代码模板/networkDelayTime.py
# Python
# LeetCode 743
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        dist = [1e9] * (n + 1)
        dist[k] = 0
        for iteration in range(n - 1):
            updated = False
            for i in range(len(times)):
                x = times[i][0]
                y = times[i][1]
                z = times[i][2]
                if dist[y] > dist[x] + z:
                    dist[y] = dist[x] + z
                    updated = True
            if not updated:
                break
        ans = 0
        for i in range(1, n + 1):
            ans = max(ans, dist[i])
        if ans == 1e9:
            ans = -1
        return ans

File: geekbang/algorithm-models/7.1- Bellman-Ford 求最短路代码模板/networkDelayTime.cpp
// C/C++
// LeetCode 743
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<int> dist(n + 1, 1e9);
        dist[k] = 0;
        for (int iteration = 1; iteration < n; iteration++) {
            bool updated = false;
            for (int i = 0; i < times.size(); i++) {
                int x = times[i][0];
                int y = times[i][1];
                int z = times[i][2];
                if (dist[y] > dist[x] + z) {
                    dist[y] = dist[x] + z;
                    updated = true;
                }
            }
            if (!updated) break;
        }
        int ans = 0;
        for (int i = 1; i <= n; i++)
            ans = max(ans, dist[i]);
        if (ans == 1e9) ans = -1;
        return ans;
    }
};
